oss-internship-2020/libtiff/test/raw_decode.cc

```diff
@@ -0,0 +1,244 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <array>
+#include <cstring>
+
+#include "helper.h"  // NOLINT(build/include)
+#include "tiffio.h"  // NOLINT(build/include)
+
+namespace {
+
+using ::sapi::IsOk;
+using ::testing::Eq;
+using ::testing::IsFalse;
+using ::testing::IsTrue;
+using ::testing::NotNull;
+
+struct ChannelLimits {
+  uint8_t min_red;
+  uint8_t max_red;
+  uint8_t min_green;
+  uint8_t max_green;
+  uint8_t min_blue;
+  uint8_t max_blue;
+  uint8_t min_alpha;
+  uint8_t max_alpha;
+};
+
+constexpr uint32_t kRawTileNumber = 9;
+constexpr uint32_t kClusterSize = 6;
+constexpr uint32_t kChannelsInPixel = 3;
+constexpr uint32_t kTestCount = 3;
+constexpr uint32_t kImageSize = 128 * 128;
+constexpr uint32_t kClusterImageSize = 64 * 64;
+using ClusterData = std::array<uint8_t, kClusterSize>;
+
+constexpr std::array<std::pair<uint32_t, ClusterData>, kTestCount> kClusters = {
+    {{0, {0, 0, 2, 0, 138, 139}},
+     {64, {0, 0, 9, 6, 134, 119}},
+     {128, {44, 40, 63, 59, 230, 95}}}};
+
+constexpr std::array<std::pair<uint32_t, ChannelLimits>, kTestCount> kLimits = {
+    {{0, {15, 18, 0, 0, 18, 41, 255, 255}},
+     {64, {0, 0, 0, 0, 0, 2, 255, 255}},
+     {512, {5, 6, 34, 36, 182, 196, 255, 255}}}};
+
+bool CheckCluster(uint32_t cluster, const sapi::v::Array<uint8_t>& buffer,
+                  const ClusterData& expected_cluster) {
+  bool is_overrun = (buffer.GetSize() <= cluster * kClusterSize);
+  EXPECT_THAT(is_overrun, IsFalse()) << "Overrun";
+
+  if (is_overrun) {
+    return true;
+  }
+
+  auto* target = buffer.GetData() + cluster * kClusterSize;
+  bool comp =
+      !(std::memcmp(target, expected_cluster.data(), kClusterSize) == 0);
+
+  // the image is split on 6-bit clusters because it has YCbCr color format
+  EXPECT_THAT(comp, IsFalse())
+      << "Cluster " << cluster << " did not match expected results.\n"
+      << "Expect: " << expected_cluster[0] << "\t" << expected_cluster[1]
+      << "\t" << expected_cluster[2] << "\t" << expected_cluster[3] << "\t"
+      << expected_cluster[4] << "\t" << expected_cluster[5] << "\n"
+      << "Got: " << target[0] << "\t" << target[1] << "\t" << target[2] << "\t"
+      << target[3] << "\t" << target[4] << "\t" << target[5];
+
+  return comp;
+}
+
+bool CheckRgbPixel(uint32_t pixel, const ChannelLimits& limits,
+                   const sapi::v::Array<uint8_t>& buffer) {
+  bool is_overrun = (buffer.GetSize() <= pixel * kChannelsInPixel);
+  EXPECT_THAT(is_overrun, IsFalse()) << "Overrun";
+
+  if (is_overrun) {
+    return true;
+  }
+
+  auto* rgb = buffer.GetData() + pixel * kChannelsInPixel;
+  bool comp = !(rgb[0] >= limits.min_red && rgb[0] <= limits.max_red &&
+                rgb[1] >= limits.min_green && rgb[1] <= limits.max_green &&
+                rgb[2] >= limits.min_blue && rgb[2] <= limits.max_blue);
+
+  EXPECT_THAT(comp, IsFalse())
```

Comment:
A series of 
```c++
EXPECT_THAT(rgb[0], Ge(limits.min_red));
EXPECT_THAT(rgb[0], Le(limits.max_red));
```
will yield shorter and clearer failure messages.

Reply:
It was all about returning bool from functions. But I'd rather not return anything then.